include/boost/geometry/algorithms/detail/is_valid/polyhedral_surface.hpp

```diff
@@ -0,0 +1,501 @@
+                        set_vec_from_point::apply(*pit2, v1);
+                        set_vec_from_point::apply(*pit3, v2);
+                        return true;
+                    }
+                }
+            }
+        }
+        return false; // All points are collinear
+    }
+
+    static point_2d project_point_to_plane_2d(vec3 const& v1, vec3 const& v2, vec3 const& p)
+    {
+        return point_2d{qvm::dot(v1, p), qvm::dot(v2, p)};
+    }
+
+    template <typename VisitPolicy>
+    static bool has_correct_intersection_orientation(polygon_2d const& polygon,
+                                                     point_2d const& point1,
+                                                     point_2d const& point2,
+                                                     VisitPolicy& visitor)
+    {
+        auto ring = boost::geometry::exterior_ring(polygon);
+        if (boost::geometry::closure<ring_2d>::value == boost::geometry::closed)
+        {
+            ring.pop_back(); // Remove the closing point to make it open
+        }
+
+        for (auto it = boost::begin(ring); it != boost::end(ring); ++it)
+        {
+            if (boost::geometry::equals(*it, point1))
+            {
+                if (boost::geometry::equals(point1, point2))
+                {
+                    visitor.template apply<failure_duplicate_points>();
+                    return false;
+                }
+                auto prev = (it == boost::begin(ring)) ? boost::end(ring) - 1 : boost::prior(it);
+                if (boost::geometry::equals(point2, *prev))
+                {
+                    // valid intersection and consistent orientation
+                    return true;
+                }
+                auto next = (it == boost::end(ring)) ? boost::begin(ring) : boost::next(it);
+                if (boost::geometry::equals(point2, *next))
+                {
+                    // valid intersection but inconsistent orientation
+                    visitor.template apply<failure_inconsistent_orientation>();
+                    return false;
+                }
+            }
+        }
+        visitor.template apply<failure_invalid_intersection>();
+        return false;
+    }
+
+    // Check if a segment intersects a polygon in 3D space
+    template <typename VisitPolicy, typename Strategy>
+    intersection_type check_polygon_segment_intersection(point_3d const& point1,
+                                                         point_3d const& point2,
+                                                         polygon_3d const& polygon,
+                                                         size_t const& polygon_index,
+                                                         VisitPolicy& visitor,
+                                                         Strategy const& strategy)
+    {
+        vec3 p1;
+        vec3 p2;
+        set_vec_from_point::apply(point1, p1);
+        set_vec_from_point::apply(point2, p2);
```

Comment:
Why not `v1` and `v2` here, if these are vectors? 
And then below you have another set of `v1` and `v2`, it's not clear to me...